server/src/main/java/org/elasticsearch/node/NodeConstruction.java

```diff
@@ -706,6 +708,8 @@ private void construct(
 
         FeatureService featureService = new FeatureService(pluginsService.loadServiceProviders(FeatureSpecification.class));
 
+        DynamicContextDataProvider.setDataProviders(pluginsService.loadServiceProviders(LoggingDataProvider.class));
```

Comment:
This seems like an arbitrary place to set the providers. Could it be done right after initializing plugins, line ~253 after createEnvironment?

Reply:
It was definitely arbitrary. I'll move it.